Unintended HUGE refactoring of automotive scanners.

polybassa opened the pull request:
This PR:
  - changes variable names of automotive utilities to be more compliant
    with python
  - adds typing for various automotive utilities
  - refactors the core of automotive scanners which involve:
	- Ecu objects
	- EcuState objects
	- EcuAnsweringMachines
	- AutomotiveTestCases
	- AutomotiveTestCaseExecutors
  - cleans up UDS Scanner, OBD Scanner and GMLAN Scanner


Sorry that this PR became that big. I hope it is merge-able in some way.

gpotter2 commented on doc/scapy/layers/automotive.rst:
```diff
@@ -913,47 +913,47 @@ Usage example:
 This utility depends heavily on the support of the used protocol. ``UDS`` is supported.
 
-Log all commands applied to an ECU
+Log all commands applied to an Ecu
```
When using abbreviations, it usually makes more sense to keep it all capitalized. Why are you changing this?

polybassa replied:
I treat `Ecu` as an utility object. I don't wanted to have it capitalized to be easier distinguishable from Layers

gpotter2 replied:
That makes sense. It might take a bit of time to review the whole PR 😄

polybassa replied:
Yes... I'm really sorry. It started small... then it grew up.

guedou commented:
I think that it is impossible to review as it is as it introduces many changes that are different. According to the PR description, there are several different changes. Do you think that submitting them in different PR will make the whole changes easier to rewview?

polybassa commented:
Unfortunately, most changes are inter-connected with each other. 

![classes](https://user-images.githubusercontent.com/1676055/105039527-f3e28100-5a60-11eb-8dd8-30529d44ee32.png)

Here is a little overview.

Besides renaming, I introduced breaking cleanups. 

The core of this scanners are the `EcuState` objects. This is an object, which stores the state of an Ecu in a generic way. 
In the `UDS` and `GMLAN` protocol, I've add knowledge about state changing packets. This means, I can determine, if a packet 
changed the state of an Ecu. 
The `Ecu` object relies on that `EcuState` to serve it's utility functions. Also the `EcuAnsweringMachine` uses this mechanisms of the `EcuState` combined with the possibility to determine state changes based on transmitted packets. This allows an `EcuAnsweringMachine` to mimic the behavior of a real Engine Control Unit. 
To configure an `EcuAnsweringMachine`, I need `EcuResponse` objects, which allow me to describe the response behavior of an `EcuAnsweringMachine` based on its current `EcuState`.

And then... I have Scanners... :-( 
Since I now have a new way to describe the internal `EcuState`, also the scanners were cleaned up an refactored to become much more generic by the use of less code.
A Scanner aka `AutomotiveTestCaseExecutor` preserves a system state `Graph` of `EcuState` objects. An  `AutomotiveTestCaseExecutor` maintains the connection to an Ecu under test and is in charge of manipulating an Ecu in a way, that all internal `AutomotiveTestCases` are executed for each available `EcuState` in the system state `Graph`. `AutomotiveTestCases` are in charge of performing tests on a target in a certain `EcuState`. A test cases sends request, in my case `UDS`, `GMLAN` or `OBD`, to a target, and evaluates the responses. As soon as a test case determines a state change of the `EcuState` of the real Ecu under test, (again from the functions implemented in the protocol), it will insert a new identified `EcuState` into the system state `Graph` of the `AutomotiveTestCaseExecutor`. 

After some tests on the old implementation, I discovered, that the implementation of the `EcuState` object was not flexible enough.. so I had to change this object, which ended up in a huge chain of further changes. On that way of changes, I also heavily cleaned up the code an introduced typing. This is, how I ended up with this unfortunately huge PR. 

I already removed unrelated changes to that core in #3036, #3035, #3034, #3033.

Since every thing is bound to this `EcuState` object, in some way, it's very hard to split this PR up. I will double check, if I can remove some minor changes which are not entangled with `EcuState`.

polybassa commented:
@guedou I've started a series of smaller PRs to make this one somehow merge-able.
